Add Password Policy API updates for OIE (#1907)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/reference/api/oie-policy-types/index.md

  }
}
```

## Updates to Password Policy

The Password Policy API went through a few changes as part of the Identity Engine. The section below shows only the delta from the [Password Policy API](/docs/reference/api/policy/#password-policy)

Earlier, we supported setting recovery factors on the policy. While we will still continue to support it for backwards compatibility of older clients, the new password policy model does not contain recovery factors. This is now moved to the password policy rule model.

### Password Policy Settings Example

```json
"settings": {
     "password": {
       "complexity": {
         "minLength": 8,
         "minLowerCase": 1,
         "minUpperCase": 1,
         "minNumber": 1,
         "minSymbol": 0,
         "excludeUsername": true
       },
       "age": {
         "maxAgeDays": 0,
         "expireWarnDays": 0,
         "minAgeMinutes": 0,
         "historyCount": 0
       },
       "lockout": {
         "maxAttempts": 10,
         "autoUnlockMinutes": 0,
         "showLockoutFailures": false
       }
     },
     "delegation": {
       "options": {
         "skipUnlock": false
       }
     }
}
```

> The recovery factors are now on the password policy rule configured inside the action `selfServicePasswordReset`. Below are 3 examples of how recovery factors are configured on the rule based on admin requirements.

Reviewer: ```suggestion
Recovery Factors have been moved from the Policy Settings object to the Password Policy Rule object, specifically inside the `selfServicePasswordReset` Action. 

Below are 3 examples of how Recovery Factors can be configured based on admin requirements.
```

> ### Password Policy Rule Action Data

Reviewer: ```suggestion
### Password Policy Rule Action Examples
```

Would this be a more accurate heading?

> Admin requires end users to verify with just 1 authenticator before they can recover password. Email, Sms, Voice and Okta Verify Push can be used by end users to initiate recovery. No step up authenticators are required.

Reviewer: ```suggestion
In this example, the requirement is that end users verify with just one authenticator before they can recover their password. Email, SMS, Voice and Okta Verify Push can be used by end users to initiate recovery. We know that only one authenticator is requirement because there are no step up authenticators specified.
```

Reviewer 2: Typo in @jakubvul 's suggested change: `We know that only one authenticator is requirement` should be `We know that only one authenticator is required`.

Reviewer: Fixed.

```json
"actions": {
      "passwordChange": {
        "access": "ALLOW"
      },
      "selfServicePasswordReset": {
        "access": "ALLOW",
        "requirement": {
          "primary": {
            "methods": [
              "EMAIL",
              "SMS",
              "VOICE",
              "PUSH"
            ]
          },
          "stepUp": {
            "required": false
          }
        }
      },
      "selfServiceUnlock": {
        "access": "ALLOW"
      }
}
```

> Admin requires end users to verify 2 authenticators before they can recover password. Email and Okta Verify Push can be used by end users to initiate recovery. Security Question is required as step-up.

Reviewer: ```suggestion
In this example, the requirement is that end users verify two authenticators before they can recover their password. Only email and Okta Verify Push can be used by end users to initiate recovery. A Security Question is required as a step-up.
```

```json
"actions": {
      "passwordChange": {
        "access": "ALLOW"
      },
      "selfServicePasswordReset": {
        "access": "ALLOW",
        "requirement": {
          "primary": {
            "methods": [
              "EMAIL",
              "PUSH"
            ]
          },
          "stepUp": {
            "required": true,
            "methods": [
              "SECURITY_QUESTION"
            ]
          }
        }
      },
      "selfServiceUnlock": {
        "access": "ALLOW"
      }
}
```

> Admin requires end users to verify 2 authenticators before they can recover password. Okta Verify Push can be used by end users to initiate recovery. Any authenticator allowed for MFA/SSO is required as step-up.

Reviewer: ```suggestion
In this example, the requirement is that end users verify two authenticators before they can recover their password. Only Okta Verify Push can be used by end users to initiate recovery. Any authenticator allowed for MFA/SSO is required as step-up.
```

> Any authenticator allowed for MFA/SSO is required as step-up

I'm not sure what this means

Reviewer 2: Maybe something like `A step-up verification is required, for which any enrolled authenticator that can be used for sign-on can be used` makes it more clear?  @jakubvul @AjayThomas-okta

Reviewer 3: Or maybe  `A step-up verification is required, for which any authenticator that can be used for sign-on are available options.`

        "access": "ALLOW"
      },
      "selfServicePasswordReset": {
        "access": "ALLOW",
        "requirement": {
          "primary": {
            "methods": [
              "PUSH"
            ]
          },
          "stepUp": {
            "required": true
          }
        }
      },
      "selfServiceUnlock": {
        "access": "ALLOW"
      }
}
```

### Self Service Password Reset Action object

| Property                      | Type    | Description                                                                                 | Supported Values
| ---                           | ---     | ---                                                                                         | ---
| `access`                      | String  | Indicates if the action is permitted                                                         | `ALLOW`, `DENY`
| `requirement`                 | Object  | JSON object containing authenticator methods required to be verified if `access` is `ALLOW` | N/A
| `requirement.primary.methods` | Array   | Authenticator methods that can be used by end user to initiate password recovery             | `EMAIL`, `SMS`, `VOICE`, `PUSH`

> | `requirement.stepUp.required` | Boolean | If any stepUp verification is required to recover password following primary methods verification | `true`, `false`

Reviewer: ```suggestion
| `requirement.stepUp.required` | Boolean | Indicates if any step-up verification is required to recover password following primary methods verification | `true`, `false`
```

> | `requirement.stepUp.methods`  | Array   | If `requirement.stepUp.required` is true, JSON object containing authenticator methods required to be verified as a stepUp. If not specified, any MFA allowed authenticator methods can be used as stepUp. | `SECURITY_QUESTION`

Reviewer: ```suggestion
| `requirement.stepUp.methods`  | Array   | If `requirement.stepUp.required` is true, JSON object containing authenticator methods required to be verified as a step up. If not specified, any MFA allowed authenticator methods can be used as step up. | `SECURITY_QUESTION`
```

> any MFA allowed authenticator methods

Not sure what this means.

Reviewer 2: Similar to above, something like `any enrolled authenticator allowed for sign-on` might be more clear.
Also, do we hyphenate "step-up" or not?  We should pick a style and stick to it.

  }
}
```

## Updates to Password Policy

> This updated Password Policy API shows only the delta from the original [Password Policy API](/docs/reference/api/policy/#password-policy)

Reviewer: ```suggestion
This updated Password Policy API shows only the changes from the original [Password Policy API](/docs/reference/api/policy/#password-policy)
```

> Earlier, we supported setting recovery factors on the [policy](/docs/reference/api/policy/#policy-settings-example-2). While we will still continue to support it for backwards compatibility of older clients, the new password policy model does not contain recovery factors. This is now moved to the password policy rule model.

Reviewer: ```suggestion
Earlier, we supported setting recovery Factors on the [Policy](/docs/reference/api/policy/#policy-settings-example-2). While we will still continue to support it for backwards compatibility of older clients, the new Password Policy model does not contain recovery Factors. This is now moved to the password policy Rule model.
```

> ### Password Policy Settings Example

Reviewer: ```suggestion
### Password Policy Settings example
```

```json
"settings": {
     "password": {
       "complexity": {
         "minLength": 8,
         "minLowerCase": 1,
         "minUpperCase": 1,
         "minNumber": 1,
         "minSymbol": 0,
         "excludeUsername": true
       },
       "age": {
         "maxAgeDays": 0,
         "expireWarnDays": 0,
         "minAgeMinutes": 0,
         "historyCount": 0
       },
       "lockout": {
         "maxAttempts": 10,
         "autoUnlockMinutes": 0,
         "showLockoutFailures": false
       }
     },
     "delegation": {
       "options": {
         "skipUnlock": false
       }
     }
}
```

Recovery Factors have been moved from the Policy Settings object to the Password Policy Rule object, specifically inside the `selfServicePasswordReset` Action.

> Below are 3 examples of how recovery factors are configured on the rule based on admin requirements.

Reviewer: ```suggestion
Below are 3 examples of how Recovery Factors are configured in the Rule based on admin requirements.
```

> ### Password Policy Rule Action Examples

Reviewer: ```suggestion
### Password Policy Rule Action examples
```

> In this example, the requirement is that end users verify with just one authenticator before they can recover their password. Email, SMS, Voice or Okta Verify Push can be used by end users to initiate recovery. We know that only one authenticator is required because there are no step up authenticators specified as can be seen by `stepUp` object containing the `required` attribute set as `false`.

Reviewer: ```suggestion
In this example, the requirement is that end users verify with just one Authenticator before they can recover their password. Email, SMS, Voice or Okta Verify Push can be used by end users to initiate recovery. We know that only one Authenticator is required because there are no step up Authenticators specified as can be seen by the `stepUp` object having the `required` attribute set as `false`.
```

```json
"actions": {
      "passwordChange": {
        "access": "ALLOW"
      },
      "selfServicePasswordReset": {
        "access": "ALLOW",
        "requirement": {
          "primary": {
            "methods": [
              "EMAIL",
              "SMS",
              "VOICE",
              "PUSH"
            ]
          },
          "stepUp": {
            "required": false
          }
        }
      },
      "selfServiceUnlock": {
        "access": "ALLOW"
      }
}
```

> In this example, the requirement is that end users verify two authenticators before they can recover their password. Only email or Okta Verify Push can be used by end users to initiate recovery. A Security Question is required as a step up.

Reviewer: ```suggestion
In this example, the requirement is that end users verify two Authenticators before they can recover their password. Only email or Okta Verify Push can be used by end users to initiate recovery. A security question is required as a step up.
```

```json
"actions": {
      "passwordChange": {
        "access": "ALLOW"
      },
      "selfServicePasswordReset": {
        "access": "ALLOW",
        "requirement": {
          "primary": {
            "methods": [
              "EMAIL",
              "PUSH"
            ]
          },
          "stepUp": {
            "required": true,
            "methods": [
              "SECURITY_QUESTION"
            ]
          }
        }
      },
      "selfServiceUnlock": {
        "access": "ALLOW"
      }
}
```

> In the final example, admin requires end users to verify 2 authenticators before they can recover password. Only Okta Verify Push can be used by end users to initiate recovery. A step up verification is required, for which any enrolled authenticator that can be used for sign-on are available options. This is indicated by the `stepUp` object containing only the `required` attribute set as `true` but without the `methods` array attribute.

Reviewer: ```suggestion
In the final example, end users are required to verify two Authenticators before they can recover their password. Only Okta Verify Push can be used by end users to initiate recovery. A step up verification is required, for which they can use any enrolled Authenticator that can be used for sign-on. This is indicated by the `stepUp` object containing only the `required` attribute set as `true` but without the `methods` array attribute.
```

```json
"actions": {
      "passwordChange": {
        "access": "ALLOW"
      },
      "selfServicePasswordReset": {
        "access": "ALLOW",
        "requirement": {
          "primary": {
            "methods": [
              "PUSH"
            ]
          },
          "stepUp": {
            "required": true
          }
        }
      },
      "selfServiceUnlock": {
        "access": "ALLOW"
      }
}
```

### Self Service Password Reset Action object

| Property                      | Type    | Description                                                                                 | Supported Values
| ---                           | ---     | ---                                                                                         | ---
| `access`                      | String  | Indicates if the action is permitted                                                         | `ALLOW`, `DENY`

> | `requirement`                 | Object  | JSON object containing authenticator methods required to be verified if `access` is `ALLOW` | N/A

Reviewer: ```suggestion
| `requirement`                 | Object  | JSON object containing Authenticator methods required to be verified if `access` is `ALLOW` | N/A
```

| `requirement.primary.methods` | Array   | Authenticator methods that can be used by end user to initiate password recovery             | `EMAIL`, `SMS`, `VOICE`, `PUSH`

> | `requirement.stepUp.required` | Boolean | Indicates if any step up verificatio is required to recover password following primary methods verification | `true`, `false`

Reviewer: ```suggestion
| `requirement.stepUp.required` | Boolean | Indicates if any step up verification is required to recover password following primary methods verification | `true`, `false`
```

> | `requirement.stepUp.methods`  | Array   | If `requirement.stepUp.required` is true, JSON object containing authenticator methods required to be verified as a step up. If not specified, any enrolled authenticator methods allowed for sign-on can be used as step up. | `SECURITY_QUESTION`

Reviewer: ```suggestion
| `requirement.stepUp.methods`  | Array   | If `requirement.stepUp.required` is `true`, JSON object containing authenticator methods required to be verified as a step up. If not specified, any enrolled Authenticator methods allowed for sign-on can be used as step up. | `SECURITY_QUESTION`
```
